Populate coding coventions (#262)
kartoza/SAEOSS-Portal: docs/src/developer/manual/coding-conventions.md

        """Set the path for the keyword database (sqlite).

        The file will be used to search for keywords for non local datasets.

        :param path: A valid path to a sqlite database. The database does
            not need to exist already, but the user should be able to write
            to the path provided.
        :type path: str

        :returns: Flag indicating if the path was set successfully.
        :rtype: boolean
        """
        self.keyword_db_path = str(path)
```

Another example:

```
class MyObject(object):

   """My new class."""

    def add_layers(scenario_dir, paths):
        """Add the layers described in a scenario file to QGIS.

        :param scenario_dir: Base directory to find path.
        :type scenario_dir: str

        :param paths: Path of scenario file (or a list of paths).
        :type paths: str, list

        :raises: Exception, TypeError, FileNotFoundError

        .. note::
            * Exception - occurs when paths have illegal extension
            * TypeError - occurs when paths is not string or list
            * FileNotFoundError - occurs when file not found
    """
```

Note the following in the above examples:

> * The first line of a docstring should be a precis of the class/method/function

Reviewer: ```suggestion
* The first line of a docstring should be a precise description of the class/method/function
```

>   expressed in less than 80 chars, terminated with a full stop and exclude

Reviewer: ```suggestion
  expressed in less than 80 chars, terminated with a full stop and excluding
```

  redundant phrases such as 'Class to do x' or 'This method does...'.
* There should be an empty line following the first docstring line.
* More detailed explanation and usage examples can follow this first line. The
  detailed explanation should not repeat the information provided in the
  parameters and returns sections.
* A line break should follow the optional detailed description.
* **param** and **type** are grouped together with no line break between them.
* If the param description is more than one line, indent the successive lines
  with 4 spaces.
* A newline should be placed after each type and rtype.
* If multiple types are allowed, separate them with commas e.g. ``:rtype: str,
  boolean``.
* If a function or method returns nothing, no **returns** section is used.
* If a function or method does not raise anything explicitly, no raises section
  is used.
* If a function or method is extremely obvious there is no need to have
  anything more than a single line docstring.
* If a function or method returns a tuple it should be be documented as
  ``:rtype: (<type>, <type>, ..)`` e.g. ``:rtype: (int, int)``.

Please also see the **api documentation how-to** section for more

> information on how to document your code properly.

Reviewer: Which document/section of documentation does this refer to?

We should be able to make this a functional link

* Favour interpolation over concatenation. For example this is **bad**:

```python
    world = 'World'
    foo = 'Hello ' + world
```

And this is **good**:

```
    world = 'World'
    food = 'Hello %s' % world
```

* Use parenthesis for long strings. For example this is **bad**:

```python
    foo = 'The quick brown fox jumps over the lazy dog. ' +
          'The slow fat rat runs around the mouldy cheese.'
```

And this is **good**:

```python
    bar = (
        'The quick brown fox jumps over the lazy dog. '
        'The slow fat rat runs around the mouldy cheese.')
```

**Note**: The good example above follows the 'pull left' principle.

* All strings should be internationalisation enabled. Please see [i18n]
  for details.
* When using gettext, alias the uggettext as tr, and do not use the common
  convention of ``_('foo')`` as the underscore trips up some tools like pylint,
  sphinx. Also using ``tr`` makes it easy to migrate code to and from Qt's
  translation system and gettext. **Note**: gettext use is deprecated in InaSAFE.

* If you use a literal string or expression in more than one place, refactor
  it into a function or variable.

> # Standard headers

Reviewer: ```suggestion
## Standard headers
```

Each source file should include a standard header containing copyright,
authorship and version metadata as shown in the exampled below.

**Example standard header**

```
# -*- coding: utf-8 -*-
"""One line description.

.. tip::
   Detailed multi-paragraph description...
"""

import os  # python core imports first
import qgis.core  # then external imports
import safe.utils.gis  # then project imports (always using full path)

__copyright__ = "Copyright 2016, The InaSAFE Project"
__license__ = "GPL version 3"
__email__ = "info@inasafe.org"
__revision__ = '$Format:%H$'

```

**Note**:: Please see [faq_developer] for details on how the revision tag
is replaced with the SHA1 for the file when the release packages are made.

> # Qt Guidelines

Reviewer: ```suggestion
## Qt Guidelines
```

        :param theIndex: Passed by the signal that triggers this slot.
        :type theIndex: int
        """
        layerId = self.polygon_layers_combo.itemData(
            theIndex, QtCore.Qt.UserRole)
        return layer_id
```

Failure to do this may result in the slot being called multiple times per event
which is usually undesirable.

Also in some cases using the Qt API will lead you into conflict with our PEP8
naming conventions for methods and variables. This is unavoidable but should
be used only in these specific instances e.g.:

```python
    def on_foo_indexChanged():
        pass
```

Qt's naming convention causes a bit of a clash when using with 'normal' python
underscore names. For this reason we adopt the following strategy:

* in designer use underscore based naming for objects
* in your concrete implementations you should be able to then use mostly
  underscore separated names except in cases where using autoconnect slots.
* in designer you should call the form a name ending in Base e.g.
  ``FooDialogBase``. By convention the concrete implementation is called the
  same sans the Base suffix e.g. ``FooDialog``.

> # Code statistics

Reviewer: ```suggestion
## Code statistics
```

* https://www.ohloh.net/p/inasafe/analyses/latest
* https://github.com/AIFDR/inasafe/network
* https://github.com/AIFDR/inasafe/graphs

> # Working with GIT

Reviewer: ```suggestion
## Working with GIT
```

* Additions to the develop branch should be made via the GitHub **pull request** mechanism
* Pull requests should preferably be **squashed** into a single commit before applying (see http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit)
* Commits and pull requests should reference the issue number they close or contribute to

> # Landscape

Reviewer: ```suggestion
## Landscape
```
